### PR TITLE
[Aptos Framework] Optional aggregator from/to int conversions

### PR DESCRIPTION
**Pull request by georgemitenkov**
### Description

**First PR in supporting aggregatable total supply in Move**

Adds `switch` function to `OptionalAggregator` that flips its underlying implementation from non-parallelizable into parallelizable and vice versa. This helps to solve upgradability into aggregator later. Also added more documentation to the code.

Important: both `new` and `switch` should not be public!

### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3048)

### CODE REVIEW

File: aptos-move/framework/aptos-stdlib/sources/optional_aggregator.move [hunk 1]
```diff
@@ -71,6 +76,48 @@ module aptos_std::optional_aggregator {
...
+
+    /// Switches between parallelizable and non-parallelizable implementations, setting
+    /// the value of the new optional aggregator to zero.
+    fun switch_and_zero_out(optional_aggregator: OptionalAggregator): OptionalAggregator {
+        let OptionalAggregator { aggregator, integer } = optional_aggregator;
+
+        if (option::is_some(&aggregator)) {
+            // In this case we convert from Some(Agg), None to None, Some(Int).
+            // First, get the limit and destroy old aggregator/integer pair.
+            let limit = aggregator::limit(option::borrow(&aggregator));
```
Comment (gelash) [inline aptos-move/framework/aptos-stdlib/sources/optional_aggregator.move:96]: maybe we can have helper functions that destroy some(aggregator), none(integer), return limit and destroy none(aggregator), some(integer), return limit. Can be re-used in destroy function at well.

Alternatively, switching shouldn't happen very often, so if we don't care about hyper-efficiency, we could just re-use destroy function here. We could also have a separate function that gets the limit and call it before line 93. (the overhead is conditional in both limit and destroy).

**Review by gelash — APPROVED** [review]
lgtm, but wait for another approval before land.

**Comment by georgemitenkov** [comment]
FYI @gelash I updated `switch` with a better version

File: aptos-move/framework/aptos-stdlib/sources/optional_aggregator.move [hunk 2]
```diff
@@ -71,18 +76,77 @@ module aptos_std::optional_aggregator {
...
+        let limit = aggregator::limit(&aggregator);
+        aggregator::destroy(aggregator);
+        let integer = new_integer(limit);
+        option::fill(&mut optional_aggregator.integer, integer);
+        limit
+    }
+
+    /// Switches from non-parallelizable to parallelizable implementation, zero-initializing
+    /// the value.
+    fun switch_and_zero_out_non_parallelizable(
```
Comment (zekun000) [inline aptos-move/framework/aptos-stdlib/sources/optional_aggregator.move:111]: I'd call it switch_to_integer_and_zero_out ,switch_to_aggregator_and_zero_out

File: aptos-move/framework/aptos-stdlib/sources/optional_aggregator.move [hunk 3]
```diff
@@ -71,18 +76,77 @@ module aptos_std::optional_aggregator {
...
+            destroy_parallelizable(optional_aggregator);
         } else {
-            destroy_integer(option::destroy_some(integer));
-            option::destroy_none(aggregator);
+            destroy_non_parallelizable(optional_aggregator);
         }
     }
 
+    /// Destroys parallelizable optional aggregator and returns its limit.
+    fun destroy_parallelizable(optional_aggregator: OptionalAggregator): u128 {
```
Comment (zekun000) [inline aptos-move/framework/aptos-stdlib/sources/optional_aggregator.move:132]: this too, destroy_aggregator/destroy_integer

**Review by zekun000 — APPROVED** [review]

**Review by movekevin — APPROVED** [review]

**Pull request merged**